src/twisted/internet/asyncioreactor.py

```diff
@@ -9,42 +9,61 @@
 
 import errno
 
+from typing import Dict, Optional, Type
+
 from zope.interface import implementer
 
 from twisted.logger import Logger
 from twisted.internet.posixbase import (PosixReactorBase, _NO_FILEDESC,
                                         _ContinuousPolling)
 from twisted.python.log import callWithLogger
 from twisted.python.runtime import seconds as runtimeSeconds
+from twisted.internet.abstract import FileDescriptor
 from twisted.internet.interfaces import IReactorFDSet
 
-try:
-    from asyncio import get_event_loop
-except ImportError:
-    raise ImportError("Requires asyncio.")
-
-# As per ImportError above, this module is never imported on python 2, but
-# pyflakes still runs on python 2, so let's tell it where the errors come from.
-from builtins import PermissionError, BrokenPipeError
+from asyncio import get_event_loop, AbstractEventLoop, SelectorEventLoop
 
 
 
 @implementer(IReactorFDSet)
 class AsyncioSelectorReactor(PosixReactorBase):
     """
     Reactor running on top of L{asyncio.SelectorEventLoop}.
+
+    On POSIX platforms, the default event loop is
+    L{asyncio.SelectorEventLoop}.
+    On Windows, the default event loop on Python 3.7 and older
+    is L{asyncio.WindowsSelectorEventLoop}, but on Python 3.8 and newer
+    the default event loop is L{asyncio.WindowsProactorEventLoop} which
+    is incompatible with L{AsyncioSelectorReactor}.
+    Applications that use L{AsyncioSelectorReactor} on Windows
+    with Python 3.8+ must call
+    C{asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())}
+    before instantiating and running L{AsyncioSelectorReactor}.
     """
     _asyncClosed = False
     _log = Logger()
 
-    def __init__(self, eventloop=None):
-
+    def __init__(self, eventloop: Optional[SelectorEventLoop] = None):
         if eventloop is None:
-            eventloop = get_event_loop()
+            _eventloop = get_event_loop()  # type: AbstractEventLoop
+            # On Python 3.8+, asyncio.get_event_loop() on
+            # Windows was changed to return a ProactorEventLoop
+            if not isinstance(_eventloop, SelectorEventLoop):
+                raise ValueError(
+                    "get_event_loop() returns: {} which is not a "
+                    "SelectorEventLoop".format(_eventloop))
+
+            eventloop = _eventloop
+        else:
+            if not isinstance(eventloop, SelectorEventLoop):
+                raise ValueError(
+                    "eventloop is: {} which is not a subclass of "
+                    "SelectorEventLoop".format(eventloop))
```

Comment:
maybe simplify the sanity checks 
 
```suggestion
            eventloop = _eventloop
 
        # On Python 3.8+, asyncio.get_event_loop() on
        # Windows was changed to return a ProactorEventLoop
        if not isinstance(eventloop, SelectorEventLoop):
            raise ValueError(
                "eventloop is: {} which is not a subclass of "
                "SelectorEventLoop".format(eventloop))
```

Reply:
Done